File: core/image/json_builder.py
"""
JSON 构建器 v2.3.2
block_grid -> living-the-grid 兼容 JSON。
本地生成：source="ns_auto_paint"。仅被使用的颜色写入 palette。
"""

from core.image.hsb_mapper import rgb_to_steps
from core.models.canvas_mode import get_canvas_mode

JSON_PRESET = {"standard": "square", "book": "book", "tv": "tv", "game": "videogame", "decoration": "interior"}


def grid_to_json(grid, palette, canvas_mode, brush_type, brush_size, is_local=False):
    mode = get_canvas_mode(canvas_mode)
    rows = len(grid)
    cols = len(grid[0]) if rows > 0 else 0

    def find_closest(r, g, b):
        best = 0
        best_d = float("inf")
        for idx, prgb in enumerate(palette):
            d = (r - prgb[0]) ** 2 + (g - prgb[1]) ** 2 + (b - prgb[2]) ** 2
            if d < best_d:
                best_d = d
                best = idx
        return best

    # 构建完整 pixels（用完整 palette 索引）
    full_pixels = []
    used = set()
    for row in grid:
        pixel_row = []
        for cell in row:
            if cell.get("a", 0) >= 128:
                idx = find_closest(cell["r"], cell["g"], cell["b"])
                pixel_row.append(idx)
                used.add(idx)
            else:
                pixel_row.append(None)
        full_pixels.append(pixel_row)

    # 仅被使用的颜色重映射
    sorted_used = sorted(used) if used else []
    idx_map = {old: new for new, old in enumerate(sorted_used)}

    jpal = []
    for ui in sorted_used:
        rgb = palette[ui]
        hs, ss, bs = rgb_to_steps(*rgb)
        if hs == 200:
            hs = 199
        elif hs == 201:
            hs = 0
        jpal.append({"hex": f"#{rgb[0]:02X}{rgb[1]:02X}{rgb[2]:02X}", "rgb": list(rgb), "press": {"h": hs, "s": ss, "b": bs}})

    pixels = []
    for row in full_pixels:
        pixel_row = []
        for idx in row:
            pixel_row.append(idx_map[idx] if idx is not None else None)
        pixels.append(pixel_row)

    return {"source": "ns_auto_paint" if is_local else "NS Auto Painter v2.3.2", "version": 2,
            "width": cols, "height": rows,
            "brush": {"mode": brush_type, "px": brush_size},
            "canvas": {"preset": JSON_PRESET.get(canvas_mode, "square"), "w": mode.active_w, "h": mode.active_h},
            "palette": jpal, "pixels": pixels}

File: frontend/src/engine/json-builder.ts
import { rgbToSteps } from './hsb-mapper';
import { getCanvasMode } from './canvas-mode';
import { GridCell, PixelJson, Palette, PaletteEntry } from './types';

const JSON_PRESET: Record<string, string> = {
  standard: 'square', book: 'book', tv: 'tv',
  game: 'videogame', decoration: 'interior',
};

/**
 * 将像素化 + 量化的 grid/matrix/palette 构建为 living-the-grid 兼容 JSON。
 */
export function gridToJson(
  grid: GridCell[][],
  palette: Palette,
  canvasMode: string,
  brushType: string,
  brushSize: number,
): PixelJson {
  const mode = getCanvasMode(canvasMode);
  const rows = grid.length;
  const cols = grid[0]?.length ?? 0;

  // 查找每个 grid cell 最近的调色板颜色
  function findClosest(r: number, g: number, b: number): number {
    let best = 0;
    let bestD = Infinity;
    for (let i = 0; i < palette.length; i++) {
      const [pr, pg, pb] = palette[i];
      const d = (r - pr) ** 2 + (g - pg) ** 2 + (b - pb) ** 2;
      if (d < bestD) { bestD = d; best = i; }
    }
    return best;
  }

  // 构建像素索引矩阵
  const fullPixels: (number | null)[][] = [];
  const used = new Set<number>();

  for (const row of grid) {
    const pixelRow: (number | null)[] = [];
    for (const cell of row) {
      if (cell.a >= 128) {
        const idx = findClosest(cell.r, cell.g, cell.b);
        pixelRow.push(idx);
        used.add(idx);
      } else {
        pixelRow.push(null);
      }
    }
    fullPixels.push(pixelRow);
  }

  // 仅包含使用的颜色，重映射索引
  const sortedUsed = [...used].sort((a, b) => a - b);
  const idxMap = new Map<number, number>();
  sortedUsed.forEach((oldIdx, newIdx) => idxMap.set(oldIdx, newIdx));

  // 构建调色板（含 press 数据）
  const jpal: PaletteEntry[] = sortedUsed.map(ui => {
    const [r, g, b] = palette[ui];
    let [hs, ss, bs] = rgbToSteps(r, g, b);
    if (hs === 200) hs = 199;
    else if (hs === 201) hs = 0;
    return {
      hex: `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}`.toUpperCase(),
      rgb: [r, g, b],
      press: { h: hs, s: ss, b: bs },
    };
  });

  // 重映射像素
  const pixels: (number | null)[][] = fullPixels.map(row =>
    row.map(idx => (idx !== null ? idxMap.get(idx)! : null)),
  );

  return {
    source: 'ns_auto_paint',
    version: 2,
    width: cols,
    height: rows,
    brush: { mode: brushType, px: brushSize },
    canvas: {
      preset: JSON_PRESET[canvasMode] || 'square',
      w: mode.activeW,
      h: mode.activeH,
    },
    palette: jpal,
    pixels,
  };
}
